### ISSUE TITLE
[UX] Add image browser in the WYSIWYG 'Insert Image' dialog.

### ISSUE THREAD

**Issue by Graham-72** [backdrop/backdrop-issues#3134]
I am aiming to provide a simple addition to core's filter module to offer the selection of an image from those already stored in the website. I am basing this on work done by @daggerhart in response to issue #1307, published as PR 1889.

I will provide a fresh PR as a proof of concept, for team review, particularly of usability.

**Update:** My PR 2198 is now working and does, I believe, provide a useful addition to the Insert Image process, particularly for non-technical users.

---
PR by @Graham-72: https://github.com/backdrop/backdrop/pull/2198

**Comment by Graham-72** [comment]
I have my initial PR (number 2198) now working in the sandbox but only by adding contrib module 'autologout' to the sandbox! I have yet to discover why this makes a difference but discovered the need for it, or something, because I found my PR worked on one of my test sites where this is installed.

Without this, there is some interaction of scripts that halts the process.

So there are now two things to do:
1. Get this to work without the supporting(?) contrib module.
2. Test and revise the UI in the light of comments and experience.

**Comment by olafgrabienski** [comment]
The PR sandbox works for me but I found the help text in the dialog (`Remember to set the image style at the foot of this dialogue before selecting Insert.`) confusing:

- I was asking myself what is 'image style' (and guessed Image size and alignment but that's not clear in my opinion.
- Scanning the text and reading the word "select", I thought that I had to set the 'image style' before selecting the *image*, which is not the case. Maybe choose another verb for "select"?
- Do we need the help text at all?

**Comment by Graham-72** [comment]
@olafgrabienski Thank you for your comment. The problem I foresee, particularly for non-technical users, is that when the library is showing several images, the fields for setting image width and alignment are not always visible unless one goes to the foot ot the dialogue, and one may miss these important settings.

I think it would be possible to place these above the images which might perhaps be better.

**Comment by Graham-72** [comment]
I don't like to use the colloquial word 'click' so what should one say (rather than 'select') about using the 'Insert' button?

**Comment by docwilmot** [comment]
@Graham-72 I thought that a better option, instead of embedding the gallery in the image dialog, we could put a button that opens a new dialog with the gallery in it. I actually tried to do this, but had an issue with the galley submit button giving a POST error. Was hoping @quicksketch could help, posted in gitter, waiting a response.

**Comment by Graham-72** [comment]
@docwilmot yes that might well be better, particularly if the gallery could give some better user feedback that a particular image has been selected.

Whilst on the subject of possible improvements, I think it would be nice to include an autocomplete for anyone keying in a filename, though image files can have very unhelpful filenames.

**Comment by Graham-72** [comment]
In the light of comments received (thank you) I have made some revisions to my PR for a version which embeds the image display in the image dialog. 

I am particularly keen to get some UI like this that enables a user to see and re-use images that have already been uploaded to a site, so I will continue work on this.

**Comment by olafgrabienski** [comment]
@Graham-72 I've had a look at the sandbox site of https://github.com/backdrop/backdrop/pull/2198 where I couldn't find any option to select an existing image in the CKEditor image dialog.

**Comment by Graham-72** [comment]
@olafgrabienski I think there are some possibilities of the image library not appearing and maybe the script needs some refinement, but when in the 'Image File URL' dialog you should see the existing images displayed below the subheading 'Select from image library' as in this screenshot:
![screenshot-2018-6-6 edit page test page 2 pr 2198 backdrop backdrop testing site](https://user-images.githubusercontent.com/10559274/41034586-efe56414-6981-11e8-8f63-22106e3c20f2.png)

**Comment by Graham-72** [comment]
@olafgrabienski could it possibly be that your PC has not yet refreshed its copy of the filter.js script?

**Comment by olafgrabienski** [comment]
Thanks for the hint @Graham-72 , that was apparently the case. I've tried it in another browser and could see the existing images.

Just one first note for the moment: While I could choose from and insert one of the existing images, I was not able to use the image library pager. When I chose the next page, the link opened in a separate normal page, not within the image dialog.

**Comment by Graham-72** [comment]
@olafgrabienski thanks for the feedback. I think the paging problem will only be solved if we can get ajax working for the view.

**Comment by laryn** [comment]
[cross-posted on the PR accidentally]

@Graham-72 Looking good! A few suggestions for consideration:

- How do you feel about using a square "scale and crop" style so all image previews are the same aspect ratio and we can get a nicely aligned rows and columns?
- If the width of `.library-view` is not hard-coded at 600px and we change from a grid display format, it would allow wider screens to display more images without requiring a scroll.
- Having a simple exposed filter for `filename contains` would be nice.

**Comment by bjoern-st** [comment]
@Graham-72 Great Work!
Just a question: how about an additional (text)field where you can set the margin around the image?

**Comment by docwilmot** [comment]
> How do you feel about using a square "scale and crop" style so all image previews are the same aspect ratio and we can get a nicely aligned rows and columns?

Unless we can show previews in the dialog, we probably shouldn't do this. A content editor might need to know the image aspect ratio when choosing an image.

> If the width of .library-view is not hard-coded at 600px and we change from a grid display format, it would allow wider screens to display more images without requiring a scroll.

I dont get what this means.

**Comment by laryn** [comment]
@docwilmot Screenshots attached.

**Current PR** (width hard-coded to 600px):
![screen shot 2018-06-08 at 10 54 02 am](https://user-images.githubusercontent.com/871421/41168408-7b476176-6b0b-11e8-8f7e-974eccbaa670.jpg)

**Slight change** (setting width to auto):
![screen shot 2018-06-08 at 10 54 24 am](https://user-images.githubusercontent.com/871421/41168425-8a042d2a-6b0b-11e8-9c6e-6a1fe285a3c0.jpg)

**Greater change** (changing view to unformatted list instead of grid and using css to display with flex, etc.):
![screen shot 2018-06-08 at 11 01 47 am](https://user-images.githubusercontent.com/871421/41168457-a48a232a-6b0b-11e8-8b1e-154abd472e74.jpg)

**Comment by Graham-72** [comment]
@laryn I like the change to use a view that is an 'unformatted list' rather than a grid. But I am not keen on using a 'scale and crop' style because I like to have a true preview of the image aspect ratio.

**Comment by laryn** [comment]
Chacun à son goût. Let me know if you want me to turn the rough into a more formal submit to your PR or if you plan to just work from what you have.

**Comment by olafgrabienski** [comment]
@Graham-72 Unfortunately I've got little time these days but here are some ideas on the fly:

- Maybe change the tab name `Image file URL` to something that makes more clear that there is an image library?
- On the image library: What about displaying square images and mentioning the aspect ratio as text?

**Comment by Graham-72** [comment]
@jenlampton please have a look at the latest version of this PR and let me know what you think from a usability view?

**Comment by Graham-72** [comment]
This PR goes some way towards adding the functionality for images that Wordpress users enjoy 'out of the box' with their 'Add Media' feature.

**Comment by jenlampton** [comment]
@Graham-72 I *love* this. It's a huge improvement over what we have now, and a killer new feature :) I was able to use it intuitively (without reading this issue first) and figure it out. 

My only thought was that maybe the form fields should be on the right and images on the left, since what I'm most excited about (as a user) was browsing the images, so I wanted to see those first. 

Also, the sort order of the view should probably have the newest images at the top, rather than at the bottom.

**Comment by docwilmot** [comment]
So after the meeting today it seems this has technical approval then @Graham-72, so full speed ahead. 👍 
Would appreciate really one of those exhaustive @klonos reviews next.

**Comment by klonos** [comment]
I would like to add to @jenlampton's comment... this is absolutely brilliant and would a kick-@ss feature to have for 1.11!! ...although I would love it even more if this was a "Media library" and not just an "Image library" ...but baby steps 😄

Comments for improvements:

- add spacing between the "Image library" label and the list of images
- the help text below the "Add a caption" checkbox looks like it needs spacing too
- make the library wrap under the settings on mobile/narrow screens
- (perhaps a separate issue for this) add some shadow on the `.ui-dialog-buttonpane` bit so that there is some visual clue that there's more things if the user scrolls further down. Similar to what @laryn has done for the form submit buttons in https://github.com/backdrop/backdrop-issues/issues/566#issuecomment-384853607 :

    ![screen shot 2018-04-26 at 10 25 10 pm](https://user-images.githubusercontent.com/871421/39343135-e9350008-49a0-11e8-8b2f-439c0cc98ce7.jpg)

- add indication of which image is selected. Perhaps a tick overlaid in the top-left/right corner of the image, and/or some border around it, and/or bolder filename text. We could aim for as much consistency as possible with what we have in `/admin/structure/layouts/settings`, since it seems like a very similar UI pattern:

    <img width="1071" alt="screen shot 2018-06-22 at 8 12 05 am" src="https://user-images.githubusercontent.com/2423362/41748223-46a3376c-75f4-11e8-8da4-d6e27453e5b7.png">

- the number of items per row should be consistent. we should be truncating long filenames (perhaps expanding on hover, but in a way that still retains layout of the rest of the files). I really like how the respective media module library lays the images out consistently:

    ![image](https://user-images.githubusercontent.com/2423362/41748976-7f1cffda-75f7-11e8-8f60-fcc5c9317d33.png)

    ...if we cannot get consensus on cropping, then perhaps consider shrink-to-fit?

- search/sort functionality should be available for the library items.

**Comment by klonos** [comment]
...thank you so much for working on this @Graham-72 ...and you too @docwilmot for the persistent feedback/help! Great team work 👍

**Comment by Graham-72** [comment]
@klonos thanks for your suggestions. Regarding your two 'add spacing' suggestions I would find it helpful if you could explain your problem a bit more. What screen size were you using? Was this when using the admin theme (Seven) or the site theme (Basis), or perhaps both? Is this a matter of personal preference or is there a usability issue? Thanks for any clarification you can provide.

**Comment by klonos** [comment]
@Graham-72 this is stock screenshot from https://www.drupal.org/project/media

**Comment by Graham-72** [comment]
@klonos 
> this was a stock screenshot from https://www.drupal.org/project/media

Thank you for this information. It means I can try the D7 version and see for myself the dynamics of this UI. At first I thought it might be something just from D8 which I do not use.

**Comment by Graham-72** [comment]
I very much like your suggestion of using a similar approach to that used in template selection based on radio buttons, and I will be trying this.

**Comment by docwilmot** [comment]
@Graham-72 noted the View uses full sized images. We should use image style thumbnails.
Is work still going on on this? Any help needed?

**Comment by Graham-72** [comment]
@docwilmot 
>We should use image style thumbnails.
I would very much like to find a way of doing this and I think we made need to revise the way the image library view is generated, because at the moment it is not treating them as images, just as files, albeit filtered by mime type, hence the thumbnails are not being created.

I have a number of other changes in the pipeline and hope to post a package of modifications to my PR soon. I will think some more about using thumbnails but would appreciate any suggestions you may have as to how to achieve this.

**Comment by Graham-72** [comment]
**Progress report:**

Thank you for the comments and suggestions received, and also those made at the weekly meeting on 28th June 2018. I have been working on some changes and have just updated the PR to include some amendments. Please kindly have another look at the sandbox site.

There are some features not yet included and maybe never will be until we have a more comprehensive media facility. For example: search for a file name; control of image margins. 

Three things that seem to me to be important for a good immediate solution are

1. A revised way of creating the image library with Views, and this requires a new Views handler as proposed by @docwilmot in issue #3183, in order that thumbnails of image files may be used rather than the original image file. Perhaps this might also allow the use of Ajax and paging, not presently possible.
2. Better truncation of long file names by omitting characters from the middle of long file names rather than the end.
3. A configuration setting so that the site designer can choose whether images are scaled to fit within the library cells (as at present) or cropped and fill the whole cell as some people prefer.

Changes that I have included in this latest update are

1. Image library now on left of screen when displayed.
2. Images sorted by View so that the most recent is first.
3. Clearer indication of which image is currently selected.
4. Styles for hovering and selecting are set in the style sheet not in javascript.
5. Tweaks to spacing of elements of dialog.

**Comment by Graham-72** [comment]
>give up on Views, at least till #3183 can be done

I think this might be a bit drastic! I think our main need is to be able to create the library view from thumbnail images rather than the full image file every time. Could we not create a folder of thumbnail images, perhaps on first use, and modify the existing view to use the thumbnail copy of the image?

For a large site with,say, hundreds of images, the use of AJAX and paging would become important, but perhaps better provided with a more advanced media option.

How difficult will #3183 be?

**Comment by klonos** [comment]
@docwilmot I agree, but it depends on what @Graham-72:

> How difficult will #3183 be?

**Comment by docwilmot** [comment]
> How difficult will #3183 be?

I responded there. It might be complex, or not, but the issue is do we want this issue held up by that one. We're almost there with this, and I'm sure we could do a themed display with FormAPI and AJAX quick time. 

> Could we not create a folder of thumbnail images

You mean we'd have to batch convert every image on every site to thumbnail as part of the upgrade path then create a thumbnail for every image thereafter. That would be inefficient to say the least.

Lets see what anyone else has to say about this.

**Comment by klonos** [comment]
The main reason for using a view is so that admins can customise it. It seems to me that if we implemented this as a view, we would:

a) be reusing parts of core. This means added features over time get automatically inherited
b) be porting any required missing features from D7 contrib modules (namely file_entity) and merging them into core. These features will be used here specifically, but will also be solving other pains elsewhere...

https://github.com/backdrop/backdrop-issues/issues/2920#issuecomment-401383857 :

> ...there is a solution suggested there, but it requires the https://drupal.org/project/file_entity module. According to https://backdrop-ops.github.io/slides/intro-mini.html#/19 this module has already been merged in core, but I do not see any "Rendered file" option for the file formatter in views:

**Comment by Graham-72** [comment]
> convert every image on every site to thumbnail

This has raised two thoughts in my mind.
1. Is this not going to happen one way or another if we want the library view to be based on downloading thumbnails rather than full image files? The objective is that this generation of thumbnails should only happen once.

2. At the moment the PR only provides for images that are in the 'inline-images' folder (or whatever has been set for this.) A typical site may contain many other images, added as specific node fields and having their own specific directory, named in the relevant field.instance config I believe. Would it not be a useful extension of the PR to provide for selection of one of the site's other image directories to be viewed in the library view? Though presumably the result of selecting and inserting such an image might have to be to copy the image to the inline-images folder? (Or is this opening a can of worms?)

**Comment by docwilmot** [comment]
Actually imagecache (image styles) does the resizing automatically. It only generates the thumbnail when you try to view it. Say you visit `path-to-thumbnail.jpg` and it doesnt yet exist, it is then created. So no we wouldnt need to batch convert if we used styles.

> provide for selection of one of the site's other image directories 

We would need a settings page for the library then. And have the View select from allowed folders. How? I dont know yet, but I suspect more custom handlers would be needed.

Note please that I am not against having a View. Was just wondering since our View limits us now, if we could temporarily do the easy thing. I've requested to discuss the File Views image rendering part of things on Thursday, so at least we can have some clarity on how much work will need to go into that issue.

**Comment by olafgrabienski** [comment]
> A configuration setting so that the site designer can choose whether images are scaled to fit within the library cells (as at present) or cropped and fill the whole cell as some people prefer.

Such a setting would indeed be helpful. I suggest to set the cropped image / filled cell as default though, that seems to be the standard display in Drupal / Media (see screenshot in https://github.com/backdrop/backdrop-issues/issues/3134#issuecomment-399269259) and in the Media Library of WordPress (see https://codex.wordpress.org/images/0/08/managefiles.png).

**Comment by klonos** [comment]
Thanks for the WP screenshot @olafgrabienski ...I don't like how one seems to need to hover over the image in order to reveal the filename, but I really like how the image name is overlaid over the image. Perhaps we would make a combination of these: the filename is shown truncated with ellipsis if too long, but when hovered on, it expands, bleeding into the image (in order to maintain thumbnail tiles dimensions at all times).

**Comment by Graham-72** [comment]
>At the moment the PR only provides for images that are in the 'inline-images' folder.

I am wrong! The view is including all images in the files directory. And those in folders other than `inline-images` will presumably already have thumbnails available on the server? Could we make use of these in the library?

**Comment by Graham-72** [comment]
>I suggest to set the cropped image / filled cell as default 

To date I have been very unenthusiastic about the cropped image / filled cell view because I think it is much more useful for a user to have a clear presentation of what the image really looks like in order to conform to the spirit of WYSIWYG. I think this could only be achieved by adding more attachment detail to the menu as done in Wordpress's Add Media selection. Screenshot:
![screenshot-2018-7-6 edit page oliverhosting wordpress](https://user-images.githubusercontent.com/10559274/42367342-c44b5160-80fb-11e8-8416-1ad5b1dd6107.png)

Even there, I am not enthusiastic because the 'true' preview is not shown until the image is preselected, so it means one or more extra steps in choosing an image. I have yet to see how D7's media module handles this. 

The type of user I have in mind is one working in a small office, who is not an expert in web design or Backdrop, but may well be tasked with editing and creating pages on a website, and also with producing MailChimp newsletters. MC has an image library and displays stored images in their correct proportion, though in my view has other UI challenges.

**Comment by olafgrabienski** [comment]
> MC has an image library and displays stored images in their correct proportion

Interesting, can you provide a screenshot of the Mailchimp display?

**Comment by Graham-72** [comment]
>provide a screenshot of the Mailchimp display

Here is the top part of one I occasionally have to use
![screenshot-2018-7-6 campaign builder - template designer mailchimp](https://user-images.githubusercontent.com/10559274/42371941-ce559302-8108-11e8-801f-3f2920951de3.png)

This screenshot is of the grid view. One can toggle to a list view that gives file name, size tec.

The MC system stores duplicates of images when different sizes are used in campaigns and the disadvantage of this is that where different size files exist one has to toggle to the list view to see the sizes of the options. For example, the second and third omages on the first line in this screenshot represent very different files, one being 34KB the other 3.6MB.

**Comment by klonos** [comment]
OK, after seeing the MC screenshot, I think that it would be OK if the image *tiles* were the same size. What MC does to achieve this is to add equal margins above/below the full, non-cropped thumbnail (or to the left/right of the thumbnail if it is in portrait).

**Comment by klonos** [comment]
...small UX nit: you know how it gets annoying when you have a toggle that gets moved to a different place once you have clicked on it? Well, when the library was to the right of the controls, this did not happen. Now it that we have moved it to the left, it does happen when you try to toggle between upload vs select from library. That's because when the dialog initially loads, it is aligned to the left of the screen. Switch it to be aligned to the right, and this issue goes away 😉

**Comment by Graham-72** [comment]
>...small UX nit: you know how it gets annoying when you have a toggle that gets moved to a different place

@klonos I am not fully understanding you because I am always seeing 'image upload' to the left of 'select from library' which is one of the toggles here. Or are you commenting on the fact that the dialogue is moving to the right of the screen?

**Comment by Graham-72** [comment]
>I think that it would be OK if the image tiles were the same size. 

@klonos again I am not understanding the point you are making. Both library displays (this PR and MC) are using fixed-size cells for their thumbnails and reducing the images proportionately to fit within these cells.

**Comment by Graham-72** [comment]
>Actually imagecache (image styles) does the resizing automatically. It only generates the thumbnail when you try to view it. Say you visit path-to-thumbnail.jpg and it doesnt yet exist, it is then created. So no we wouldnt need to batch convert if we used styles.

@docwilmot thanks for this comment four days ago. If only I could modify the file path token [uri] used by the view in the first of its two global custom text fields by inserting '/styles/thumbnail/public' into the file URI it would use thumbnails rather than the original image files.

This would be simple if I could use some PHP here 😄  I have attempted to use a template to modify the output for this field but have had no success. Any ideas?

**Comment by klonos** [comment]
> @klonos I am not fully understanding you...

Ok, screenshot time! 😄 ...

![image](https://user-images.githubusercontent.com/2423362/42423720-a84d3516-8342-11e8-9a1e-8c8bb7e77081.png)

So I have intentionally grabbed this ^^ on a 20" monitor. Notice how:

a) the dialog is aligned to the left of the screen when it is loaded.
b) when I click on the "Select from library" link, it expands and the controls are not to the right side of the screen.
c) my mouse cursor pauses half-way as it travels across the other end of the screen (trying to reach the controls). This is because I have to *physically* lift my mouse up, move it back, then continue the mouse moving.

Now, if the dialog was popping up to the right of the screen, then image gallery would be the one to expand to the left, and my mouse would still be at the portion of the screen that holds the controls:

![image](https://user-images.githubusercontent.com/2423362/42423752-f6c69cc8-8342-11e8-90e8-54b0431137d3.png)

...see how I barely have to move my mouse now?

**Comment by klonos** [comment]
![42371941-ce559302-8108-11e8-801f-3f2920951de3](https://user-images.githubusercontent.com/2423362/42423820-0cea3c20-8344-11e8-8192-81693e1558b7.png)

Notice how both portrait and landscape images have equal paddings above/below and left/right?

Well, our version does this only for portrait images, and top-aligns landscape ones:

![screen shot 2018-07-09 at 6 52 16 am](https://user-images.githubusercontent.com/2423362/42423870-3c3a0e00-8345-11e8-93d8-08d38aa85183.png)

I find the MC version better because it also adds a subtle gray background color to fill the padding. This makes all the image thumbnails look square, whereas we can only tell that these are tiles when hovered or clicked on.

**Comment by Graham-72** [comment]
@klonos thank you for your detailed and very well illustrated explanations to which I respond as follows.

Regarding the style of the image gallery, it seems that in summary there are only two differences you are suggesting
(a) for images that do not fill their cell vertically, to equalise the space top and bottom, rather than align their tops.
(b) to provide a light background color for every cell as in MC, not just on hover or selection.
These are fine adjustments to the CSS which can easily be implemented if this is the concensus.

Regarding the position of the pointer I have a very different view which I will try to explain. First I would point out that although there is a form of toggle action, this is not a true toggle between 'Image library' and 'Select from library' because they are separated physically, as in the Insert Link dialogue where mouse movement is required from one to another. More importantly, if one considers a users sequence of actions, the normal flow is surely having clicked 'Select from library' to click on one of the image thumbnails, and the pointer is already within that screen area. This may be happy coincidence but to me it seems right and there would be no justification in adding to the script in order to move the pointer over to the right.

**Comment by Graham-72** [comment]
I have made some more modifications to the PR so that the Image Library is now using thumbnail image files created by a new Image Style I have named 'thumbnail_150' which scales image to a maximum height of 150px for further scaling if necessary by CSS to fit within the image library cell 150px square.

To modify the Library Image view I have added a template file in the templates folder of core's theme Seven. I am not sure that this is the right thing to do, but it works!

**Comment by olafgrabienski** [comment]
@Graham-72 Thanks for the Mailchimp screenshot.

> (suggestions by @klonos) :
(a) for images that do not fill their cell vertically, to equalise the space top and bottom, rather than align their tops.
(b) to provide a light background color for every cell as in MC, not just on hover or selection.
These are fine adjustments to the CSS which can easily be implemented if this is the concensus.

Personally, I still prefer the cropped square images because they look more nicely but I see that the 'scaled to fit' approach like in Mailchimp has other advantages. (Still hoping that the respective configuration setting as suggested in https://github.com/backdrop/backdrop-issues/issues/3134#issuecomment-402020604 will be available.) I agree with both of @klonos' suggestions regarding the 'scaled to fit' approach and would like to add another one for a better look and feel: remove the borders from, and add some margin between the 'cells'.

Another topic: testing the PR on the sandbox site I didn't find an easy way to deselect a selected image using the mouse and the keyboard (tried it with click + `cmd` on a Mac).

**Comment by Graham-72** [comment]
Thanks for comments.

I have been thinking more about this and personally concluded that in order to achieve an aim of making this very familiar to my users, it would be best to make this image library very similar in style and operation (i.e. selection of an image) to what the majority of them are used to, which I suggest is File Explorer in Windows 10 in default style. See this partial screenshot:
![win10_file_explorer](https://user-images.githubusercontent.com/10559274/42766633-d81000dc-8912-11e8-9db3-f7d56c945f3c.jpg)

The thumbnails have a white background unless hovered over when it becomes pale blue with no border, and a slightly stronger blue with a thin border when selected. There is no provision for 'unselecting' except by selecting a different one. The thumbnail images are proportional to their originals. They are centered horizontally in their grid cells but vertically aligned at the bottom.

This would then be consistent with what they see when selecting and uploading a single image file from their PC, using the Insert Image / Choose file procedure.

**Comment by Graham-72** [comment]
I have now revised the PR so that the Image Library View matches the styles seen in Windows 10 File Explorer and also uses the standard thumbnail image style used in Backdrop (100 x 100 px).

**Comment by olafgrabienski** [comment]
> making this very familiar to my users, it would be best to make this image library very similar in style and operation (i.e. selection of an image) to what the majority of them are used to, which I suggest is File Explorer in Windows 10 in default style

In my opinion, the Windows style isn't visually attractive. I see the point of many people being used to Windows 10 but I wouldn't choose the design of a specific operating system for the default but the look and feel of the Backdrop user interface. @Graham-72 - what do you think of providing the Windows 10 orientated design as one of several configurations to choose from?

**Comment by klonos** [comment]
Not that Windows Explorer should be setting the example of best UX (😛...this coming from a Windows user ...that has turned Mac for the sake of work ...that does not particularly like it ...and that secretly wants to switch to Linux when the time is right). Having said that, I do agree that the majority of people globally are using Windows.

For completeness, and in order to get a better idea of commonly-used patterns, here's a screenshot of how things look in Mac Finder:

![screen shot 2018-07-19 at 2 06 43 am](https://user-images.githubusercontent.com/2423362/42893755-927f05f8-8af8-11e8-9156-255c77ead1e4.png)

- There's a frame added around the thumbnails (in which I personally see no real value)
- Long filenames are wrapped in a second line, but beyond that, they are shortened, with a ... somewhere in the middle of the name.
- There is a rounded, light grey shadow for the selected files, and also their filename gets a blue background.

Hare, also what the file manager in Ubuntu looks like (this one could use a little bit of ❤️TBH):

![screen shot 2018-07-19 at 8 55 01 am](https://user-images.githubusercontent.com/2423362/42971753-c7c790ce-8bf0-11e8-930c-eddc12dd2ed3.png)

**Comment by klonos** [comment]
I wanted to provide screenshots to demo how things work in WP, but at the same time did not want to derail this issue here. So: #3203

**Comment by Graham-72** [comment]
@jenlampton I am now stuck on making progress with this issue because of differing views on usability. I think moving the image library to the the left of the screen as you suggested is a big improvement but there are many options for style and differing views on this. I am reluctant to add options for the style of the library because (a) it will take more development time and (b) it results in more code (bloat?).

Is this something your Design and Usabilty meeting might advise on? I just want to achieve a clear and easy-to-use process for when a user selects the 'image' icon. I envisage that in the future there might also be a 'media' icon in the WYSIWYG editor that provides a more comprehensive set of features.

**Comment by jenlampton** [comment]
@Graham-72 I'm also of the mind that less is more here, and we should shoot for getting it done over getting it perfect. The PR [and sandbox](http://2198.backdrop.backdrop.qa.backdropcms.org) as it is now is *so much better* than what's in core now,  my recommendation would be that we get this in as-is, and then iterate if necessary.

The only significant change that I think is necessary is making the library responsive. On smaller screens let's stack the right sidebar below the thumbnails. This may not be the perfect solution, but It's good enough to get it in! The majority of the time people will be inserting images from a laptop anyway.

**Comment by Graham-72** [comment]
@jenlampton thanks for taking a look at this again. I have now modified the styles so that the screen is responsive and for widths narrower than 1030px the form is displayed below the library grid of thumbnails. I chose this break point because it is where the horizontal scroll would otherwise cut in. I have also made some small changes and additions to wording so that hopefully it is more intelligible to users.

**Comment by jenlampton** [comment]
I left one minor suggestion about the view on the PR, but otherwise this looks great to me!  Let's get her in and then iterate :)

**Comment by docwilmot** [comment]
I also left a comment. I would say though that I really prefer the gallery being on the right side so that the location of the form is consistent when you toggle.
Ideally though we need to fix the location of the dialog. It should be centered at all times rather than left-aligned for the upload toggle. Also after toggle it doesnt re-center itself properly till you scroll the whole page and I see a quick 'jump' to the center on scrolling. 
And a nice-to-have would be animation during the toggle.

**Comment by jenlampton** [comment]
Good ideas @docwilmot, can we get them into a follow up issue?

**Comment by docwilmot** [comment]
I feel that the UX issues with the form placement, and the form jumping around on scroll should be sorted out before we seal it into core. It wouldn't feel right otherwise.

**Comment by Graham-72** [comment]
I don't see it as any great problem to implement these changes if we are sure that is what we want to do, though I'm not sure about the animation on toggling.

**Comment by Graham-72** [comment]
>I left one minor suggestion about the view on the PR.

@jenlampton thank for this suggestion which I have incorporated with an adaptation that enables it to use the thumbnail image files rather than the full sized ones. I am not sure whether I have implemented this in a Backdrop-approved way! I have taken the views handler for the uri field and added an extra checkbox option so that this field can be set to provide the thumbnail url, derived from the main image url. I have added this custom views handler to the Filter module on the assumption that this means it will have no impact on views belonging to other modules.

I have also made some adjustments to the stylesheet for the dialogue form, in the light of some of the comments made.

**Comment by quicksketch** [comment]
I left a code review on the PR: https://github.com/backdrop/backdrop/pull/2198#pullrequestreview-143237948


Some issues worth pulling out from the PR:

- This removes the "Image URL" option entirely, which makes using external images difficult or impossible to use. Should we be introducing a 3rd tab instead of replacing the "Image URL" with "Select from library"?
- The option to configure the image style seems like it should be a part of the view, not a part of the text format configuration.

I also agree with all @docwilmot's suggestions above: put the gallery on the right, center the dialog, and animate the transition.

**Comment by Graham-72** [comment]
@quicksketch thanks for the code review. I have made several changes to the PR. External image URLs can now be entered instead of selecting one of the library images. The image style files used for the thumbnails is now selected within the view from a selection list of those available, mainly to ensure that a valid style name is selected. The actual size in the display is controlled by styles in filter.css.

Regarding appearance of the UI, I was strongly influenced by the comment above on 21 June by @jenlampton who wrote
>My only thought was that maybe the form fields should be on the right and images on the left, since what I'm most excited about (as a user) was browsing the images, so I wanted to see those first.

I therefore changed them, and prefer the result. However, they can of course be changed back.

For narrower screens they are now centred one above the other but I am strongly of the opinion that on wider screens it is better to make full use of the width and have them alongside each other.

I don't know about animating the transition. How exactly?

**Comment by docwilmot** [comment]
@Graham-72 the problem I have with the gallery on the left is **just** because the entire dialog is not centered. To explain: notice when you first click the image icon, the dialog consists of only the form, and it is stuck on the left of the page. It seems natural then that when you toggle the gallery, that form should stay on the left and the gallery appear on the right. It seems really weird that the form moves so dramatically from entirely on the left side of the page to entirely on the right.

This is why I mentioned that 'ideally' we should center the dialog and animate it. If on first load the dialog (just the upload form) is centered, then on toggling it wouldn't matter much which side the gallery is as it would be obvious that it would no longer be able to stay in the center. If it was animated, it would be visually more satisfying to see it move and return to the center with the toggle.

**Comment by Graham-72** [comment]
@docwilmot Thanks for the added explanation. Now I understand, and yes that would be nice.

**Comment by quicksketch** [comment]
I can push a change to the PR to animate and center the dialog. I don't think that will take much work, just a CSS transition and triggering the auto-centering code on the tab toggle.

**Comment by quicksketch** [comment]
I pushed up a commit to the current PR that animates the dialog. I also found that the `data-fid` tag was not being added to images, so I updated the click handler and the view to include the file FID.

The config file `views.fiew.image_library.json` has to be included for new installations to pick up the view, so I exported it and included it in the PR. That means we have the entire view twice: once in a config file for new installations, and once in the .install file to add it to existing sites. That's required in order to make it so that we get consistent starting points when we update the view in the future.

I also flipped the form (again) so that the fields are on the left and the library is on the right. IMO it makes for a smoother transition.

The JS might need additional cleanup, I found it a bit hard to follow, although it seems to work.

There are a few critical issues that I've found while testing:

- The view right now shows _all_images, with no pager. That's not going to work on any site with more than a dozen images.
- The AJAX behaviors still don't work within the view, so configuring a pager or exposed filters is not an option.
- Updating an existing image to replace it with a different one does not work. Selecting a replacement image works but the dialog does not save and close when the Save button is clicked.

**Comment by quicksketch** [comment]
I rebased the PR at https://github.com/backdrop/backdrop/pull/2198 as it's been open for quite a while. I think I found the AJAX issue with Views: Views was passing in a jQuery object instead of DOM Element for the "context" of the new AJAX objects. I corrected the Views AJAX instantiations but also made it so that jQuery objects will be converted back into DOM elements if necessary.

With that fixed, I updated the View to re-enable AJAX and turn back on the pager. There was also a fair amount of cruft in the view that could be removed, and I removed excess fields that weren't being displayed.

**Comment by Graham-72** [comment]
Thank you for all the changes to this PR. It is looking really good except that it now seems to be inserting the thumbnail of the image rather than the stored image file, so the necessary re-calculation of the URL has perhaps been removed?

**Comment by quicksketch** [comment]
Doh! Yes, looks like I may have broken it. I can correct that tomorrow. Can you keep looking and see if the other changes look acceptable?

**Comment by Graham-72** [comment]
If I add an image at the foot of the edit window and right align it, it does not seem possible to add anything after it because CKEditor will not allow an insertion point after the image. This seems to be the case whether or not there is a caption. I wonder whether this is a more general CKEditor problem because I have experienced something like this in other circumstances.

**Comment by quicksketch** [comment]
Thanks folks, I've updated the PR to increase the robustness of the relative-URL generation. I don't think it's entirely safe to attempt to reverse-engineer the original URL out of an image style. Some modules like CDN or S3Connector might make local URLs for image styles but remote ones for the original images. So we need to separately include both the thumbnail URL and the full URL (via a data attribute), then attempt to strip off the domain name for portability.

**Comment by herbdool** [comment]
With the latest commit the data attribute and thumbnail issues are fixed though I don't know what else needs to be tested.

**Comment by quicksketch** [comment]
Things I can think of that need validation:

- Uploading an image still works.
- Editing an image that has been uploaded opens the dialog to the upload version, with the value prepopulated so the image can be detached/removed.
- Selecting an image from the browser can insert an image.
- Selecting an image from the browser can _replace_ an image.
- Manually inserting an image (both remote and relative) URL works.
- Editing an image that is manually inserted continues saving the URL.
- Copy/paste an image still works.
- Editing an image that is copy/pasted in should work as though the image were uploaded through the dialog.
- The pager to go between multiple pages of images works.
- Images can be inserted after using the pager.

**Comment by docwilmot** [comment]
Tested all above, including drag and drop, all work normally.
One unexpected finding is that if you paste an image into CKEditor, then doubleclick to edit, the dialog toggles switch position with library toggle to the left of the upload toggle. This persists anytime you doubleclick that particular image, but when editing other images the toggles stay where they were.

FWIW I would not consider this a blocker.

**Comment by docwilmot** [comment]
But.. Couple of notices in logs:
`Notice: Undefined index: callback in ajax_form_callback() (line 394 of /home/qa/github/backdrop/backdrop/pr/2198/core/includes/ajax.inc).`

and

`Notice: Undefined index: type in filter_format_editor_image_form_submit() (line 332 of /home/qa/github/backdrop/backdrop/pr/2198/core/modules/filter/filter.pages.inc).`

**Comment by jenlampton** [comment]
I left [a comment](https://github.com/backdrop/backdrop/pull/2198#issuecomment-418000641) on the PR.

**Comment by quicksketch** [comment]
Super, thanks @jenlampton! With that, I think we've gotten this to a stable and strong state, though possibilities for improvement are still out there.

I've merged https://github.com/backdrop/backdrop/pull/2198 into 1.x for 1.11.0. Thanks everyone for all the excellent feedback. Credit is in https://github.com/backdrop/backdrop/commit/368a744e35362d09bde4c0924ff86293090fd173. Thanks especially @Graham-72 for his incredible commitment to keep coming back to this month after month! :rocket:

**Comment by klonos** [comment]
...follow-up tickets filed 😉
